synapse/appservice/api.py

```diff
@@ -388,6 +388,46 @@ async def push_bulk(
         failed_transactions_counter.labels(service.id).inc()
         return False
 
+    async def claim_client_keys(
+        self, service: "ApplicationService", query: List[Tuple[str, str, str]]
+    ) -> Tuple[Dict[str, Dict[str, Dict[str, JsonDict]]], List[Tuple[str, str, str]]]:
+        """
+        Returns:
+            A tuple of:
+                A map of user ID -> a map device ID -> a map of key ID -> JSON dict.
+
+                A copy of the input which has not been fulfilled because the
+                appservice doesn't support this endpoint..
+        """
+        if service.url is None:
+            return {}, query
+
+        # This is required by the configuration.
+        assert service.hs_token is not None
+
+        # Create the expected payload shape.
+        body: Dict[str, Dict[str, List[str]]] = {}
+        for user_id, device, algorithm in query:
+            body.setdefault(user_id, {}).setdefault(device, []).append(algorithm)
+
+        # TODO Does service.url end up with the /v1 in it?
+        uri = f"{service.url}/unstable/org.matrix.msc3983/keys/claim"
```

Comment:
`service.url` appears to be whatever is defined in the registration file. Normally that'll be `http://localhost:9000` or similar, which means this becomes `http://localhost:9000/unstable/org.matrix.msc3983/keys/claim`. 

If someone is trying to work around the Synapse issue where it doesn't append the correct path prefix by using `http://localhost:9000/_matrix/app/v1` in their config, this indeed does the `/v1/unstable` thing.

See also https://github.com/matrix-org/synapse/issues/3780

For now I'll patch over this with the bot-sdk: https://github.com/turt2live/matrix-bot-sdk/pull/307/files#diff-5298577ea3609c1f482d041a248acfbb71f2d5e6a53d7dcb248453b6d9ef86a0R297-R299

Reply:
I figured I should just fix the root of the problem see #15317. I'll update this PR to at least put the `/_matrix/app` in front for now though.

Reply:
```suggestion
        # TODO Does service.url end up with the /v1 in it?
        uri = f"{service.url}/_matrix/app/unstable/org.matrix.msc3983/keys/claim"
```